### PR TITLE
Big simplification of slabs combination

### PR DESCRIPTION
**Pull request by kilbith**
Why doing things complicated when you can make them simple ?

### CODE REVIEW

**Comment by everamzah** [comment]
Seems alright at first, but placing a slab adjacent on the side results in a combined node.  Normally, doing this would result in an extension, like a bridge.  I like it, though, it's fast.

**Comment by Jordach** [comment]
James, I think building support structures is more fun - then you've got to
plan your buildings correctly, and not relying on silly "game physics".
Scaffolding is cheap, if you have the dirt.

**Comment by kilbith** [comment]
> Seems alright at first, but placing a slab adjacent on the side results in a combined node.

Can you be clearer, please ? You mean if the first slab is placed on its side ?

**Comment by everamzah** [comment]
Okay I will test it out.  I was just thinking about how the on_place callback is important here for other reasons: such as checking protection!

**Comment by everamzah** [comment]
I will attempt to be more clear: Placing the slab on the side of another slab normally results in another slab.

With your patch:
screenshot_20160125_082507

Current behavior:
screenshot_20160125_082810

**Comment by everamzah** [comment]
Jordach does, however, bring up a very interesting point about gameplay here, and I think he's onto something.

**Comment by kilbith** [comment]
Ah, right. Well you can hold sneak + right-click for placing slabs alongside.

**Comment by everamzah** [comment]
That's true, and should be good enough.  The problem I find is that you have to press sneak+RMB to place anything at all to the slab!  Also, protection is not being checked.

**Comment by kilbith** [comment]
Protection check added.

**Comment by everamzah** [comment]
Protection check works as expected.  However, I find it awkward that one cannot place another node to a slab without holding down sneak.  Is it reasonable to check that the player isn't holding a slab of the same type?

**Comment by kilbith** [comment]
Fixed.

**Comment by paramat** [comment]
Beautifully simple but makes building slabs to slabs more difficult, builders won't accept it. So unfortunately :-1:

**Comment by kilbith** [comment]
Oh yes, pressing sneak is very difficult. Too bad missing a much faster code.

**Comment by C1ffisme** [comment]
@paramat Actually, when I'm building roofs I hold my sneak key down, since falling off would mean getting sand/dirt to get back up again.

Although I can see a problem with this if you were flying, since the sneaking and descent keys are the same key.

**Comment by everamzah** [comment]
Personally it is not a matter of having to depress an extra key, kilbith, it is more a matter of breaking expected behavior.  If you right click a node that is not a furnace or a chest you get an adjacent node.  I can already envision the first bug report entitled "can't place anything on slabs".

**Comment by kilbith** [comment]
What are you talking about ? You can place nodes on slabs currently (without sneak), check latest modifications.

**Comment by everamzah** [comment]
I didn't see your "fixed" comment, and hadn't checked before commenting.  My apologies!

I tested now and see everything Looks Good To Me!

**Comment by kilbith** [comment]
OK I've solved the slab-to-slab thing by checking the top face of the pointed node.

Now the feature works exactly than before, but is much faster.

@paramat I'd like a re-evaluation / testing, please.

**Comment by everamzah** [comment]
They're not snapping to the bottom of another slab, like the original method does.

**Comment by kilbith** [comment]
Fixed.

**Comment by everamzah** [comment]
Adding to the side of a slab attached to the ceiling places the slab below.
screenshot_20160125_233917

**Comment by everamzah** [comment]
Here, too
screenshot_20160125_234111

**Comment by paramat** [comment]
Okay seems potentially good, although i'm not familiar with current slab behaviour as i don't build much.

**Comment by kilbith** [comment]
Currently there's just one thing remaining to solve : placing slab with facedir at 20 under a regular block, still using the same callback.

Consider this PR as WIP, but is worth it. I leave it inactive for now.

**Comment by everamzah** [comment]
When this is all said and done, I think you should consider using minetest.place_node() instead of minetest.set_node(), in order to register the placement of the slab with the on_place callback.

This may remove the need for the protection check you added.

The rational is some mod checking the number of slabs placed.  I did a quick search and didn't see any old mods relying on this, offhand, but one example would be using rubenwardy's awards mod and registering an award for number of slabs placed.

**Comment by kilbith** [comment]
You're missing the essential point of using `on_rightclick`.

The current method is : placing a slab above -> removing this slab -> placing a new node below.
The new method : placing a new node at pointed position.

It's more efficient.

**Comment by everamzah** [comment]
Yes, but doing so without registering this placement may not be proper.  This was my initial thought in response to what is otherwise a very "snappy" enhancement!

Har har.

File: mods/stairs/init.lua [hunk 1]
```diff
@@ -124,87 +132,31 @@ function stairs.register_slab(subname, recipeitem, groups, images, description,
...
-				-- If the item was taken from the fake stack, decrement original
-				if success then
-					itemstack:set_count(fakestack:get_count())
-				-- Else put old node back
+			if wield_item == "stairs:slab_" .. subname and node.param2 <= 3 and face then
+				if face == 1 then
+					minetest.set_node(pos, {name = recipeitem, param2 = node.param2})
 				else
-					minetest.set_node(slabpos, slabnode)
+					minetest.set_node(pointed_thing.above, {name = wield_item, param2 = 20})
```
Comment (ShadowNinja) [inline mods/stairs/init.lua]: What if pointed_thing.above is a solid node (can happen if it's a nodebox).
You should use a helper like item_place_node for this, there are a lot of things to get wrong.

**Comment by paramat** [comment]
Since kilbith has left, anyone feel like finishing this, if indeed it is a good method?

**Comment by paramat** [comment]
Will close in 1 month, please copy the code for your use if you intend to rework this.

**Comment by sofar** [comment]
possibly related: #1109 

Maybe I'll brush this up, since #1109 exposes a simple issue and I think we should incorporate some elements of this PR in it (if not whole).

**Comment by paramat** [comment]
Replaced by #1111

**Pull request closed without merging**